### PR TITLE
Relaxed rvec utils

### LINKED ISSUE #15077: Passing different floating point types to `RVec` utility functions
### Explain what you would like to see improved and how.

(I will be using `DeltaPhi` for illustration here, but I suspect many functions have similar "limitations". The current way of doing things is perfectly valid and might even be preferable, this issue is also meant to gauge whether you would be willing to accept PR(s) with these changes in place).

The current interface of `DeltaPhi` in `RVec.hxx` looks like this:
https://github.com/root-project/root/blob/7a4af49dc65a812a3feea24c98a49b84cd85caaf/math/vecops/inc/ROOT/RVec.hxx#L2876

This means that calling it with
```cpp
auto dphi = DeltaPhi(3.14f, 1.23)
```
will lead to a template deduction error because both argument should have the same type. We have triggered this via RDataFrame from python by using the version that uses an `RVec` by passing a `vector<float>` and `vector<double>`. The error message you get at the moment tells us exactly that, but is also not the easiest to interpret if you are not too familiar with C++.

One way around this is to introduce a temporary variable that converts one of the two vectors into the same type as the other. While this obviously works it's also not really elegant to do so.

The "proper"(?) way to fix this would to make `DeltaPhi` a bit more "relaxed" with its template arguments, something along the lines of:
```cpp
template <typename T, typename U = T>
std::common_type_t<T, U> DeltaPhi(T v1, U v2, std::common_type_t<T, U> c = M_PI) { 
  // implementation unchanged apart from updated static_assert to take into account T and U
}
```
This would then have to be done also the the overloads taking an `RVec` to solve our issue.

Is this something you would be interested in at all? Then we could maybe also discuss who would be doing the actual work for this.

### ROOT version

6.18/00 onwards

### Installation method

any 

### Operating system

any

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tmadlener**
# This Pull request:

Makes it possible to call `VecOps::DeltaPhi`, `VecOps::DeltaR`, `VecOps::DeltaR2`, `VecOps::InvariantMass` and `VecOps::InvariantMasses` with different floating point types.

## Changes or fixes:

Previously all arguments where deduced to be the same types which does not allow to call these with mixed floating point type arguments. Now each argument get's its own template type, making it possible to call them with mixed floating point types. The return type and any intermediate calculations will be done by the largest necessary floating point type, as determined via `std::common_type`.

I tried to follow existing conventions, e.g. using `Common_t` as an additional defaulted template argument for a bit less typing in the implementations. This also requires at least c++14 as it uses the `common_type_t` helper type for a bit less typing.

Some of the functions now are quite heavily templated, I am not entirely sure the full combinatorial explosion is necessary for all of them.

I haven't added tests yet. There are some pre-existing where I could "attach" them. I could also just stick a bunch of `static_assert`s for the return types etc into the implementation directly if you prefer that.

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes #15077

### CODE REVIEW

**Comment by dpiparo** [comment]
Hi @tmadlener , thanks for this PR. Could you please add coverage for this feature by extending the existing RVec tests? If that makes sense for you, once this is done, I will launch the builds for testing the new code.

**Comment by tmadlener** [comment]
I have added a few tests that check he expected return types. They also give some information about the accuracy compared to a calculation done with all doubles, since there will always be some floating point promotion somewhere in mixed cases.

File: math/vecops/test/vecops_rvec.cxx [hunk 1]
```diff
@@ -1212,6 +1212,26 @@ TEST(VecOps, DeltaPhi)
    auto dphi4 = DeltaPhi(v4, v3);
    auto r4 = -1.f * r3;
    CheckEqual(dphi4, r4);
+
+   // Checks that calling with different argument types works and yields the
+   // expected return types and values
+   static_assert(std::is_same_v<decltype(DeltaPhi(0.f, 0.0)), double>,
```
Comment (dpiparo) [inline math/vecops/test/vecops_rvec.cxx:1218]: Could you use the gtest macro EXPECT_TRUE to uniform this smart check to the rest of the code?
For example
```cpp
EXPECT_TRUE(std::is_same_v<decltype(DeltaPhi(0.f, 0.0)), double>) << "DeltaPhi should return double if one of the arguments is double";
```
This applies to other places in the code.

**Review by dpiparo — CHANGES_REQUESTED** [review]
Very good coverage. Thanks for adding it. I propose to apply a minimal change, to replace the static_asserts with gtest standard macros in order to make the code uniform. 
After that, we are ready to test.

**Comment by dpiparo** [comment]
Hi,

The changes are properly picked up, but two tutorials now fail because the number of template arguments changed. Could you fix them?
```
tutorial-dataframe-df102_NanoAODDimuonAnalysis
tutorial-v7-ntuple-ntpl004_dimuon
```
Thanks!

**Comment by tmadlener** [comment]
Ah, I didn't have the patience to run all the tests locally so I missed those. I have update the definitions of the template functions to only have one non-defaulted template argument. This should make for a truly transparent change. On the other hand in deduced contexts it should now be possible to have different floating point types.

Maybe there should be an actual test that checks that when invoked from an RDataFrame from python. I could not find an immediate best place for that, but maybe you have one in mind?

**Pull request merged**